Use RHEL 9 when creating a remote source for custom file type (#3238)
theforeman/foreman-documentation: guides/common/modules/proc_creating-a-remote-source-for-a-custom-file-type-repository.adoc

.Prerequisites
ifdef::katello,orcharhino[]

> * You have a server running {EL} 9 or {EL} 8 registered to your {Project}.

Reviewer: ```suggestion
* You have a server running {EL} 9 registered to your {Project}.
```

For nightly, Foreman+Katello does not run on EL8 anymore. Foreman 3.12 is the last release that runs on EL8: https://yum.theforeman.org/releases/3.12/ "el8" is missing for "3.13" and nightly.

If you want to only add this to 3.12 and 3.11, then please open a PR against 3.12.

endif::[]
ifdef::satellite[]

> * You have a server running {EL} 9 or {EL} 8 registered to your {Project} or the Red{nbsp}Hat CDN.

Reviewer: I do not know about Satellite, but I assume that the same as above applies.

.Prerequisites
ifdef::katello,orcharhino[]

> * You have a server running {EL} 9 registered to your {Project}.

Reviewer: AFAIK, you can still run Foreman/Katello on EL8: https://yum.theforeman.org/releases/3.12/el8/x86_64/

```suggestion
* You have a server running {EL} 9 or {EL} 8 registered to your {Project}.
```

endif::[]
ifdef::satellite[]

> * You have a server running {EL} 9 registered to your {Project} or the Red{nbsp}Hat CDN.

Reviewer 2: Is there a reason why we allow users to use EL 9 or 8 in upstream but only EL 9 in Satellite?

Author: I have a question out to Ian about this.

Reviewer 3: RHEL 8 and RHEL 9 should work in all cases, not just upstream.

Reviewer 3: Actually, I'm chatting with Odilon now about where the pulp_manifest package actually lives, it may not even be in the utils repo any more. Let me confirm first.

Reviewer 3: Okay, so it turns out the repository where pulp_manifest is stored completely changed. Even for Satellite 6.15.

The repository is now the main Satellite one: satellite-6.15-for-rhel-8-x86_64-rpms

For 6.16, it'll be satellite-6.16-for-rhel-8-x86_64-rpms and satellite-6.16-for-rhel-9-x86_64-rpms

For upstream, the pulp_manifest package is available in the pulpcore repository, which will already be enabled on the Katello machine.

Reviewer 3: I verified that the docs are correct for Satellite 6.14. So it's just 6.15+ that need correction.

Author: Changed the attribute to reflect the new repository.

Reviewer 2: @bangelic Ian says that it should run on both RHEL 9 and 8. So:
```suggestion
* You have a server running {EL} 9 or {EL} 8 registered to your {Project} or the Red{nbsp}Hat CDN.
```
But this isn't the only change required. You have also provide the step to enable the repos on EL 8 below.

* Your server has an entitlement to the {RHELServer} and {ProjectName} Utils repositories.
endif::[]
* You have installed an HTTP server.
ifndef::orcharhino[]
For more information about configuring a web server, see {RHELDocsBaseURL}9/html/deploying_different_types_of_servers/setting-apache-http-server_deploying-different-types-of-servers[Setting up the Apache HTTP web server] in {RHEL}{nbsp}9 _Deploying different types of servers_.
endif::[]

> .Procedure

Author: @ianballou We can track this under another ticket after this one gets merged.

endif::[]

.Procedure
ifdef::katello[]
. On your server, enable the required repositories:
+
[options="nowrap" subs="+quotes,attributes"]
----

> # dnf config-manager --add-repo=https://yum.theforeman.org/pulpcore/nightly/el9/x86_64/

Reviewer: ```suggestion
# dnf config-manager --add-repo=https://yum.theforeman.org/pulpcore/{PulpcoreVersion}/el9/x86_64/
```

I like that we fix the procedure for upstream Foreman+Katello, but I currently see two minor issues:

* We should use the attribute that tracks the Pulpcore version. Another issue: the attribute should be "nightly" on the "master" branch in foreman-documentation -> we can fix this in a follow-up PR.
* To my knowledge, "dnf config-manager" cannot set a URL for the GPG key. -> we can convert this to a simpler "cat EOF" thingy (https://linuxize.com/post/bash-heredoc/#basic-heredoc-examples) in a follow-up PR.

Note that both issues need a special "if master" then "no GPG key and then no attribute", but we can use the `DocState` attribute for make that happen. I can provide a follow-up PR for that.

> # dnf config-manager --add-repo=https://yum.theforeman.org/pulpcore/{PulpcoreVersion}/el9/x86_64/

Reviewer: This does not work as is:

````
Public key for python3.11-pulp_manifest-3.0.0-5.el9.noarch.rpm is not installed
The downloaded packages were saved in cache until the next successful transaction.
You can remove cached packages by executing 'dnf clean packages'.
Error: GPG check FAILED
````

My reproducer:

````
$ podman run --rm -it almalinux:9.5

# dnf install -y 'dnf-command(config-manager)'
# dnf config-manager --add-repo=https://yum.theforeman.org/pulpcore/3.63/el9/x86_64/
# echo "gpgkey=https://yum.theforeman.org/pulpcore/3.63/GPG-RPM-KEY-pulpcore" >> /etc/yum.repos.d/yum.theforeman.org_pulpcore_3.63_el9_x86_64_.repo 
# dnf makecache && dnf install -y pulp-manifest
# pulp-manifest --help
````

So either we add the `echo "..." >> pulp.repo` command or use `echo` to write the whole repo file. I do not think that `dnf config-manager` to setup the URL for the GPG pub key.

Reviewer 3: I vote we add
```
# echo "gpgkey=https://yum.theforeman.org/pulpcore/3.63/GPG-RPM-KEY-pulpcore" \
>> /etc/yum.repos.d/yum.theforeman.org_pulpcore_3.63_el9_x86_64_.repo
```

That was the only extra line I needed to get it working.

Reviewer 3: (replacing the 3.63 with `{PulpcoreVersion}`)

Author: > {PulpcoreVersion}

Updated.

endif::[]

.Procedure
ifdef::katello[]
. On your server, enable the required repositories:
+
[options="nowrap" subs="+quotes,attributes"]
----
# dnf config-manager --add-repo=https://yum.theforeman.org/pulpcore/{PulpcoreVersion}/el9/x86_64/
# echo "gpgkey=https://yum.theforeman.org/pulpcore/{PulpcoreVersion}/GPG-RPM-KEY-pulpcore" \

> >> /etc/yum.repos.d/yum.theforeman.org_pulpcore_3.63_el9_x86_64_.repo

Reviewer: ```suggestion
>> /etc/yum.repos.d/yum.theforeman.org_pulpcore_{PulpcoreVersion}_el9_x86_64_.repo
```

[options="nowrap" subs="+quotes,attributes"]
----
# subscription-manager repos \
--enable={RepoRHEL9AppStream} \
--enable={RepoRHEL9BaseOS} \
--enable={RepoRHEL9ServerSatelliteUtils}
----

> . If you are on {EL} 8, enable the satellite-utils module:

Reviewer 2: ```suggestion
. If you run your web server on {EL} 8, enable the `satellite-utils` module:
```

[options="nowrap" subs="+quotes,attributes"]
----
# subscription-manager repos \
--enable={RepoRHEL9AppStream} \
--enable={RepoRHEL9BaseOS} \
--enable={RepoRHEL9ServerSatelliteUtils}
----

> . Enable the satellite-utils module:

Reviewer 4: AFAIK there's no `satellite-utils` DNF module on EL9 so I think this step became redundant.

Author: @ekohl Should the step be reverted back to what it was or is it good as is?

Author: @ekohl Following up.

Author: @ekohl Following up again to see if this needs to be reverted back to what it was before or if it is good as is?

Reviewer 5: Line 34 describes enabling a module. EL8 uses modules, EL9 doesn't. If you are documenting a procedure that is to be performed on EL9, you can drop this whole step. If you are documenting a procedure that can be performed on either EL8 or EL9, you can precede this step with something like "If you are <doing XYZ> on EL8, enable the satellite-utils module".
